get_prism_data.R

```diff
@@ -4,39 +4,232 @@ library(prism)
+#the DB, then extract values from the raw prism rasters, store in DB, and return them.
+########################################################
+get_prism_data=function(){
+  #Query the prism data and check to see if it throws an error. 
+  prism_bbs_data=try(collect(tbl(database, sql('SELECT * from prism_bbs_data'))))
+  
+  #If it works it should be of type list. If it doesn't then it should be of class 'try-error'
+  if(typeof(prism_bbs_data)=='list'){
+    #If no error, return the data. 
+    return(as.data.frame(prism_bbs_data))
+  } else if(class(prism_bbs_data)=='try-error') { 
+    #If the query returned an error, assume it was because the table doesn't exist
+    #and need's to be created/loaded from raw prism data. 
+    print('PRISM data table not found, processing raw data')
+    
+    #Load the bbs data locations and convert them to a spatial object.
+    #Stop here if bbs data isn't available. Could also make this query the DB as well.
+    bbs_data <- try(read.csv("data/bbs_data.csv"))
+    if(class(bbs_data)=='try-error'){stop("Can't load bbs_data.csv inside get_prism_data()")}
+    locations <- unique(select(bbs_data, site_id, long, lat))
+    coordinates(locations) <- c("long", "lat")
+    
+    #Check to see if all the raw data in the years specified are downloaded,
+    #download everything again if not. (Getting only what is needed, say if a 
+    #previous download failed, might be overly complicated)
+    if(!check_If_Prism_Files_Present(ls_prism_data(), yearsToUse)){
+      downloadPrism()
+    }
+    
+    #Load the prism data and extract using the bbs locations.
+    prism_stacked <- prism_stack(ls_prism_data())
+    extracted <- raster::extract(prism_stacked, locations)
+    prism_bbs_data <- data.frame(site_id = locations$site_id, coordinates(locations), extracted)
+    prism_bbs_data <- prism_bbs_data %>%
+      gather(date, value, 4:ncol(prism_bbs_data)) %>%
+      tidyr::extract(date, c("clim_var", "year", "month"),
+                     "PRISM_([:alpha:]*)_stable_[:alnum:]*_([:digit:]{4})([:digit:]{2})_")
+    
+    #Format the data a little and load into the sqlite database.
+    prism_bbs_data$year <- as.numeric(prism_bbs_data$year)
+    prism_bbs_data$month <- as.numeric(prism_bbs_data$month)
+    mydata <- copy_to(database, prism_bbs_data, temporary = FALSE,
+                      indexes = list(c("site_id", "year", "month")))
+    
+    #Now return the data as asked for
+    return(prism_bbs_data)
+    
+  } else {
+    
+    stop('Im stopping because the sqlite query for prism_bbs data returned neither a data frame nor an error in get_prism_data()')
+  }
+  
+}
+
+
+#################################################################
+#From raw prism monthly values calculate all the bioclim variables.
+#You should not call this directly to load bioclim vars. Instead call
+#get_bioclim_data(), which will 1st try to load the data from the sdqlite
+#db before processing it all from scratch. 
```

Comment:
This function documentation should be directly attached to the function and the helper function gets it own separate docs (which it basically already has).